rfcs/2020-05-25-2685-dev-workflow-simplification.md

```diff
@@ -0,0 +1,244 @@
+
+### Using Nix
+
+> **Note:** We're still new at this Nix stuff, so if you're a Nix expert do feel free to share knowledge.
+
+If you're a Nix user, or you're open to trying out a new tool, you can use our Nix expressions!
+
+If you don't have Nix yet, [install it](https://nixos.org/download.html):
+
+```bash
+curl -L [https://nixos.org/nix/install](https://nixos.org/nix/install) | sh
+# Hesitating? Uninstalling is just `rm -rf /nix`
+```
+
+Next, run `nix-shell` from the Vector directory.
+
+***Wow, you did it!*** Now you can run the commands described in "Bring your own toolbox" now. This has pulled in all the required packages and set required environment variables.
+
+Your other programs are there too, so you can run `code .` or `clion .` or whatever and get going like normal.
+
+We've only partially adopted Nix to ensure contributors can still use familiar tools. You may still need to run `bundle install`, `yarn` or other commands to initialize things. (PRs welcome if you have ideas to improve this!)
+
+If you're interested in having Nix **automatically start your environment**, you can consider a tool like [`direnv`](https://direnv.net/docs/installation.html), often available from your package manager.
+
+```bash
+❯ cd /git/timberio/vector
+❯ direnv allow .
+❯ cd ../vector
+direnv: loading /git/timberio/vector/.envrc
+direnv: using nix
+cdirenv: export +AR +AS +CC +CONFIG_SHELL +CXX +HOST_PATH +IN_NIX_SHELL +LC_ALL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_BUILD_CORES +NIX_BUILD_TOP +NIX_CC +NIX_CC_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +PROTOC +PROTOC_INCLUDE +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +TEMP +TEMPDIR +TMP +TMPDIR +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +name +nativeBuildInputs +nobuildPhase +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~PATH
+```
+
+Now you can use the jobs detailed in **"Bring your own toolbox"** below. 
+
+### Bring your own toolbox
+
+> **Targets:** This option is required for MSVC/Mac/FreeBSD toolchains.
+
+To build Vector on your own host will require a fairly complete development environment!
+
+We keep an up to date list of all dependencies used in our CI environment inside our `default.nix` file. Loosely, you'll need the following:
+
+- **To build Vector:** Have working Rustup, C++/C build tools (LLVM, GCC, or MSVC), Python, and Perl, `make` (the GNU one preferably), `bash`, `cmake`, and `autotools`
+- **To run integration tests:** Have `docker` available, or a real live version of that service.
```

Comment:
For k8s we'll also need `minikube` (alongside `docker`) to be available on the host for integration tests.
For dev flow, the `minikube` *or something like Docker Desktop* plus `skaffold` is required.

Reply:
Yeah -- I think the k8s jobs will have their own special deps :(

Reply:
Would be cool if we pull them in via nix too - at least `skaffold` and `minikube`.

Reply:
We can! https://nixos.org/nixos/packages.html?channel=nixpkgs-unstable&query=skaffold  https://nixos.org/nixos/packages.html?channel=nixpkgs-unstable&query=minikube

Reply:
We can pass the docker socket into the environment, no big deal. And that won't be an issue in the nix shell either! Only in docker :)

Reply:
I'm afraid passing docker socket is not enough, we'll also need to pass socket to the docker-in-docker that `minikube` starts on the node. It a level 2 inception, and the issue there might be not the docker socket itself (which is indeed trivial), but the overlay network address of minikube docker.
We can try to figure this out, but I don't have high hopes in this. Should work fine with nix on the host though, so "just use nix".

Reply:
It might be the best option for kube :( I wonder if we can run inside minikube like `podman` or `docker` and solve it that way?

Reply:
Inside `minikube`? Well, `minikube` has various drivers, interesting ones are `docker` (default), `none` and `podman`. `podman` is experimental though, and we won't be able to use it in the integration yet. `none` might be a good fit when running in the in-docker environment. Actually, no, it only seems like it's a good fit, but it won't work properly in docker container with docker socket from host passed inside...

Reply:
Could we spawn a container in the kube and run in there? So it'd be a knob like `CONTAINER_TOOL=podman`? You'd set `CONTAINER_TOOL=minikube` for kubernetes and get a shell into a kubernetes pod running the env with what you need?

Reply:
`minikube` itself with driver `docker` runs containers. It'll run them in `docker`. We'll then have a docker-in-docker. So, if minikube itself runs in docker, we might get a
- docker-in-docker-in-docker situation - this is hell for doing any kind of work against
- minikube in docker, but with using host docker socket - this might work, but we'll likely have issues with overlay networks being isolated - i.e. `environment` container (where `minikube` and our other stuff is running) won't have access to the network of the k8s cluster node container that `minkube` spawned.

I initially estimated the complexity of keeping `minikube` in docker-powered environment at the k8s RFC, and I think my estimate still holds. It's simply not worth it. The amount of complexity that docker adds is way too high. It's not going to be a helpful tool, but rather a constant cause of issues.
Managing the binaries with nix for use on host is already more than enough, and super cool! :smile:

Reply:
Okay, so no action right now. :)